### PR TITLE
Optimistic-locking dataset write with etag

### PR DESCRIPTION
**Pull request by 1000TurquoisePogs**
Builds upon https://github.com/zowe/zowe-common-c/pull/247 to implement writing.
Details in https://github.com/zowe/zss/pull/355

### CODE REVIEW

**Review by lchudinov — CHANGES_REQUESTED** [review]
Please see my comments on memory management.

File: c/datasetjson.c [hunk 1]
```diff
@@ -78,6 +81,153 @@ typedef struct Volser_tag {
...
+  fclose(in);
+
+  return lrecl;
+}
+
+/*
+ TODO this duplicates a lot of stremDataset. Thinking of putting conditionals on streamDataset writing to json stream, but then function becomes misleading.
+ */
+static char *getDatasetETag(char *filename, int recordLength, int *rc, int *eTagReturnLength) {
+  char *eTag;
```
Comment (adityadr-joy) [inline c/datasetjson.c:155]: eTag pointer can be initialized to NULL

File: c/datasetjson.c [hunk 2]
```diff
@@ -78,6 +81,153 @@ typedef struct Volser_tag {
...
+  else {
+    zowelog(NULL, LOG_COMP_RESTDATASET, ZOWE_LOG_DEBUG, "FAILED TO OPEN FILE\n");
+  }
+
+  if (!rcEtag) {
+    // Convert hash text to hex.
+    eTagLength = digest.hashLength*2;
+    eTag = safeMalloc(eTagLength+1, "datasetetag");
+    memset(eTag, '\0', eTagLength);
+    int len = digest.hashLength;
```
Comment (adityadr-joy) [inline c/datasetjson.c:212]: len is unused variable

File: c/datasetjson.c [hunk 3]
```diff
@@ -110,12 +270,24 @@ int streamDataset(Socket *socket, char *filename, int recordLength, jsonPrinter
...
       zowelog(NULL, LOG_COMP_RESTDATASET, ZOWE_LOG_DEBUG, "FAILED TO OPEN FILE\n");
   }
 
   jsonEndArray(jPrinter);
+
+  if (!rcEtag) {
+    // Convert hash text to hex.
+    int eTagLength = digest.hashLength*2;
+    char eTag[eTagLength+1];
+    memset(eTag, '\0', eTagLength);
```
Comment (adityadr-joy) [inline c/datasetjson.c:285]: memset length should be eTagLength+1

File: c/datasetjson.c [hunk 4]
```diff
@@ -110,12 +270,24 @@ int streamDataset(Socket *socket, char *filename, int recordLength, jsonPrinter
...
+    int len = digest.hashLength;
```
Comment (adityadr-joy) [inline c/datasetjson.c:286]: len is unused variable

**Review by lchudinov — APPROVED** [review]
It looks good to me. Please only add a NULL check for `lastEtag`

**Pull request merged**